File: Code/RolePaymentSystem2.0/src/ec/edu/espe/rolepaymentsystem/view/UserInterface.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.rolepaymentsystem.view;

import ec.edu.espe.rolepaymentsystem.controller.EmployeeManager;
import ec.edu.espe.rolepaymentsystem.controller.PayrollGenerator;
import ec.edu.espe.rolepaymentsystem.model.Employee;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Scanner;

/**
 *
 * @author Erick Tufiño
 */
public class UserInterface {
    private static final Scanner SCANNER = new Scanner(System.in);
    private final EmployeeManager employeeManager;
    private final PayrollGenerator payrollGenerator;
    private static final double BASIC_SALARY = 460.0; 

    public UserInterface(EmployeeManager employeeManager, PayrollGenerator payrollGenerator) {
        this.employeeManager = employeeManager;
        this.payrollGenerator = payrollGenerator;
    }

    public void showMainMenu() {
        boolean exit = false;
        while (!exit) {
            System.out.println("Seleccione una opción:");
            System.out.println("1. Agregar empleado");
            System.out.println("2. Ver empleados");
            System.out.println("3. Editar empleado");
            System.out.println("4. Eliminar empleado");
            System.out.println("5. Generar roles de pago");
            System.out.println("6. Salir");
            int option = SCANNER.nextInt();
            SCANNER.nextLine(); // Consumir el salto de línea

            switch (option) {
                case 1:
                    addEmployee();
                    break;
                case 2:
                    viewEmployees();
                    break;
                case 3:
                    editEmployee();
                    break;
                case 4:
                    deleteEmployee();
                    break;
                case 5:
                    payrollGenerator.generatePayrolls(employeeManager.getEmployees());
                    break;
                case 6:
                    exit = true;
                    break;
                default:
                    System.out.println("Opción inválida");
            }
        }
    }

    private void addEmployee() {
        try {
            System.out.print("Ingrese el nombre: ");
            String name = SCANNER.nextLine();

            System.out.print("Ingrese el apellido: ");
            String lastName = SCANNER.nextLine();

            System.out.print("Ingrese el número de identificación: ");
            String idNumber = SCANNER.nextLine();

            System.out.print("Ingrese la fecha de contratación (dd/MM/yyyy): ");
            String dateStr = SCANNER.nextLine();
            SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
            Date hireDate = sdf.parse(dateStr);

            System.out.print("Ingrese las horas extras: ");
            double overtimeHours = SCANNER.nextDouble();
            SCANNER.nextLine(); // Consumir el salto de línea

            System.out.print("Ingrese los días ausentes: ");
            double absentDays = SCANNER.nextDouble();
            SCANNER.nextLine(); // Consumir el salto de línea

            System.out.print("Ingrese las bonificaciones: ");
            double bonuses = SCANNER.nextDouble();
            SCANNER.nextLine(); // Consumir el salto de línea

            System.out.print("Ingrese los préstamos del IESS: ");
            double iessLoans = SCANNER.nextDouble();
            SCANNER.nextLine(); // Consumir el salto de línea

            System.out.print("Ingrese los préstamos de la empresa: ");
            double companyLoans = SCANNER.nextDouble();
            SCANNER.nextLine(); // Consumir el salto de línea

            System.out.print("Ingrese las multas: ");
            double fines = SCANNER.nextDouble();
            SCANNER.nextLine(); // Consumir el salto de línea

            System.out.print("¿El empleado trae su propia comida? (true/false): ");
            boolean bringOwnFood = SCANNER.nextBoolean();
            SCANNER.nextLine(); // Consumir el salto de línea

            Employee employee = new Employee(name, lastName, idNumber, hireDate, BASIC_SALARY, overtimeHours, absentDays, bonuses, iessLoans, companyLoans, fines, bringOwnFood);

            employeeManager.addEmployee(employee);

            System.out.println("Empleado agregado correctamente.");
        } catch (ParseException e) {
            System.out.println("Error en el formato de la fecha. Por favor, ingrésela en el formato dd/MM/yyyy.");
        } catch (Exception e) {
            System.out.println("Error al ingresar los datos. Inténtelo nuevamente.");
        }
    }

    private void viewEmployees() {
        List<Employee> employees = employeeManager.getEmployees();
        if (employees.isEmpty()) {
            System.out.println("No hay empleados registrados.");
        } else {
            System.out.println("Lista de empleados:");
            for (int i = 0; i < employees.size(); i++) {
                Employee employee = employees.get(i);
                System.out.println((i + 1) + ". " + employee.getName() + " " + employee.getLastName());
            }
        }
    }

    private void editEmployee() {
        viewEmployees();
        if (!employeeManager.getEmployees().isEmpty()) {
            try {
                System.out.print("Ingrese el número del empleado a editar: ");
                int employeeIndex = SCANNER.nextInt() - 1;
                SCANNER.nextLine();

                if (employeeIndex >= 0 && employeeIndex < employeeManager.getEmployees().size()) {
                    Employee employee = employeeManager.getEmployees().get(employeeIndex);
                    System.out.print("Ingrese el nuevo nombre (presione Enter para mantener el actual): ");
                    String newName = SCANNER.nextLine();
                    if (!newName.isEmpty()) {
                        employee.setName(newName);
                    }

                    System.out.print("Ingrese el nuevo apellido (presione Enter para mantener el actual): ");
                    String newLastName = SCANNER.nextLine();
                    if (!newLastName.isEmpty()) {
                        employee.setLastName(newLastName);
                    }

                    // Continúa solicitando y actualizando los demás campos del empleado

                    employeeManager.updateEmployee(employeeIndex, employee);
                    System.out.println("Empleado editado correctamente.");
                } else {
                    System.out.println("Número de empleado inválido.");
                }
            } catch (Exception e) {
                System.out.println("Error al editar el empleado. Inténtelo nuevamente.");
            }
        }
    }

    private void deleteEmployee() {
        viewEmployees();
        if (!employeeManager.getEmployees().isEmpty()) {
            try {
                System.out.print("Ingrese el número del empleado a eliminar: ");
                int employeeIndex = SCANNER.nextInt() - 1;
                SCANNER.nextLine(); // Consumir el salto de línea

                if (employeeIndex >= 0 && employeeIndex < employeeManager.getEmployees().size()) {
                    employeeManager.removeEmployee(employeeIndex);
                    System.out.println("Empleado eliminado correctamente.");
                } else {
                    System.out.println("Número de empleado inválido.");
                }
            } catch (Exception e) {
                System.out.println("Error al eliminar el empleado. Inténtelo nuevamente.");
            }
        }
    }
}

File: 06-Code/RolePaymentSystem_Python/view/UserInterface.py
from controller.EmployeeManager import Employee
from datetime import datetime
from controller.EmployeeManager import EmployeeManager
from controller.PayrollGenerator import PayrollGenerator

class UserInterface:
    BASIC_SALARY = 460.0

    def __init__(self, employee_manager=EmployeeManager(), payroll_generator=PayrollGenerator()):
        self.employee_manager = employee_manager
        self.payroll_generator = payroll_generator

    def show_main_menu(self):
        exit_flag = False
        while not exit_flag:
            print("Seleccione una opción:")
            print("1. Agregar empleado")
            print("2. Ver empleados")
            print("3. Editar empleado")
            print("4. Eliminar empleado")
            print("5. Generar roles de pago")
            print("6. Salir")
            option = self.get_int_input("Opción inválida\n")

            if option == 1:
                self.add_employee()
            elif option == 2:
                self.view_employees()
            elif option == 3:
                self.edit_employee()
            elif option == 4:
                self.delete_employee()
            elif option == 5:
                self.payroll_generator.generate_payrolls(self.employee_manager.get_employees())
            elif option == 6:
                exit_flag = True

    def add_employee(self):
        employees = self.employee_manager.get_employees()
        try:
            print("Ingrese el nombre: ")
            name = self.get_user_input("Nombre vacío\n")

            print("Ingrese el apellido: ")
            last_name = self.get_user_input("Apellido vacío\n")

            print("Ingrese el número de identificación: ")
            id_number = self.get_user_input("Número de identificación vacío\n")

            print("Ingrese la fecha de contratación (dd/MM/yyyy): ")
            date_str = input()
            hire_date = datetime.strptime(date_str, "%d/%m/%Y").date()

            print("Ingrese las horas extras: ")
            overtime_hours = self.get_double_input("Horas extras vacías\n")

            print("Ingrese los días ausentes: ")
            absent_days = self.get_double_input("Días ausentes vacíos\n")

            print("Ingrese las bonificaciones: ")
            bonuses = self.get_double_input("Bonificaciones vacías\n")

            print("Ingrese los préstamos del IESS: ")
            iess_loans = self.get_double_input("Préstamos al IESS vacíos\n")

            print("Ingrese los préstamos de la empresa: ")
            company_loans = self.get_double_input("Préstamo de la empresa vacío\n")

            print("Ingrese las multas: ")
            fines = self.get_double_input("Multas vacías\n")

            print("El empleado trae su propia comida? (True/False): ")
            bring_own_food = self.get_boolean_input("Entrada inválida\n")

            employee = Employee(name, last_name, id_number, hire_date, self.BASIC_SALARY, overtime_hours, absent_days, bonuses, iess_loans, company_loans, fines, bring_own_food)
            
            employee_existing = any(emp.get_id_number() == employee.get_id_number() for emp in employees)

            if employee_existing:
                print("!!ESTE EMPLEADO YA FUE INGRESADO!!")
            else:
                self.employee_manager.add_employee(employee)
                print("Empleado agregado correctamente.")
        except ValueError as e:
            print("Error en el formato de la fecha. Por favor, ingrésela en el formato dd/MM/yyyy.")
        except Exception as e:
            print("Error al ingresar los datos. Inténtelo nuevamente.")

    def view_employees(self):
        employees = self.employee_manager.get_employees()
        if not employees:
            print("No hay empleados registrados.")
        else:
            print("Lista de empleados:")
            for i, employee in enumerate(employees, 1):
                print(f"{i}. {employee.get_name()} {employee.get_last_name()}")

    def edit_employee(self):
        self.view_employees()
        if self.employee_manager.get_employees():
            try:
                print("Ingrese el número del empleado a editar: ")
                employee_index = int(input()) - 1

                if 0 <= employee_index < len(self.employee_manager.get_employees()):
                    employee = self.employee_manager.get_employees()[employee_index]
                    print("Ingrese el nuevo nombre (presione Enter para mantener el actual): ")
                    new_name = self.get_user_input("Nombre vacío\n")
                    if new_name:
                        employee.set_name(new_name)

                    print("Ingrese el nuevo apellido (presione Enter para mantener el actual): ")
                    new_last_name = self.get_user_input("Apellido vacío\n")
                    if new_last_name:
                        employee.set_last_name(new_last_name)

                    self.employee_manager.update_employee(employee_index, employee)
                    print("Empleado editado correctamente.")
                else:
                    print("Número de empleado inválido.")
            except Exception as e:
                print("Error al editar el empleado. Inténtelo nuevamente.")

    def delete_employee(self):
        self.view_employees()
        if self.employee_manager.get_employees():
            try:
                print("Ingrese el número del empleado a eliminar: ")
                employee_index = int(input()) - 1

                if 0 <= employee_index < len(self.employee_manager.get_employees()):
                    self.employee_manager.remove_employee(employee_index)
                    print("Empleado eliminado correctamente.")
                else:
                    print("Número de empleado inválido.")
            except Exception as e:
                print("Error al eliminar el empleado. Inténtelo nuevamente.")

    @staticmethod
    def get_user_input(error_message):
        try:
            return input().lower().strip()
        except Exception:
            print(error_message)
            return ""

    @staticmethod
    def get_double_input(error_message):
        try:
            return float(input())
        except ValueError:
            print(error_message)
            return 0.0

    @staticmethod
    def get_boolean_input(error_message):
        try:
            return input().lower() == "true"
        except Exception:
            print(error_message)
            return False

    @staticmethod
    def get_int_input(error_message):
        try:
            return int(input())
        except ValueError:
            print(error_message)
            return 0